File: Python/EllipticCurveCryptography.py
from EllipticCurveArithmetic import EllipticCurveArithmetic
from Helper import Helper
from Key import Key
from Point import Point

class EllipticCurveCryptography:
	'''
	__init__: Initialise classes needed for Encryption
    Inputs:-
        G: Base Point of elliptic curve
        numPoints: Number of points in Elliptic curve
        h: Instance of Helper class
        ec: Instance of EllipticCurveArithmetic class
	'''
	def __init__(self, G, numPoints, h, ec):
		self.G = G
		self.numPoints = numPoints
		self.h = h
		self.ec = ec

	'''
	generateKey: Generate Public,Private key for user
    Input:-
        len: Number of points in Elliptic curve
    Output:-
        Return generated key
	'''
	def generateKey(self, keyLen):
		privateKey = self.h.randomNumber(keyLen)
		publicKey = self.ec.multiply(self.G, privateKey, self.h)
		return Key(publicKey, privateKey)

	'''
	Encrypt: Encrypt a message using Receiver's Public Key
    Input:-
        str: Plaintext message
        public_key: Receiver's Public Key
    Output:-
        Return Encrypted message
	'''
	def Encrypt(self, plaintext, publicKey):
		msgLen = len(plaintext)
		ciphertext = ""
		for i in range(0, msgLen):
			r = self.h.randomNumber(self.numPoints)
			A = self.ec.multiply(self.G, r, self.h)
			B = self.ec.add(self.h.charToPoint(plaintext[i]),
							self.ec.multiply(publicKey, r, self.h), self.h)

			ciphertext += self.h.pointToStream(A, self.ec.getM())
			ciphertext += self.h.pointToStream(B, self.ec.getM())

		return ciphertext

	'''
	Decrypt: Decrypt the ciphertext using Receiver's Private Key
    Input:-
        str: Ciphertext
        private_key: Receiver's Private Key
    Output:-
        Return decrypted message
	'''
	def Decrypt(self, ciphertext, privateKey):
		v = self.h.streamToPoint(ciphertext, self.ec.getM())
		plaintext = ""
		for i in range(0, len(v), 2):
			c1 = v[i]
			c2 = v[i + 1]
			DM = self.ec.subtract(c2, self.ec.multiply(c1, privateKey, self.h), self.h)
			plaintext += self.h.pointToChar(DM)

		return plaintext

File: JAVA/EllipticCurveCryptography.java
package JAVA;


import java.util.Vector;

public class EllipticCurveCryptography {

    Point G;
    Long numPoints;
    Helper h;
    EllipticCurveArithmetic ec;

    /*
    EllipticCurveCryptography: Initialise classes needed for Encryption
    Inputs:-
        G: Base Point of elliptic curve
        numPoints: Number of points in Elliptic curve
        h: Instance of Helper class
        ec: Instance of EllipticCurve class
     */
    public EllipticCurveCryptography(Point G, Long numPoints, Helper h, EllipticCurveArithmetic ec){
        this.G = G;
        this.numPoints = numPoints;
        this.h = h;
        this.ec = ec;
    }

    /*
    generateKey: Generate Public,Private key for user
    Input:-
        len: Number of points in Elliptic curve
    Output:-
        Return generated key
     */
    public Key generateKey(int len){
        Point publicKey;
        Long privateKey;
        privateKey = Helper.randomNumber((long)len);
        publicKey = ec.multiply(G, privateKey, h);
        return new Key(publicKey,privateKey);
    }

    /*
    Encrypt: Encrypt a message using Receiver's Public Key
    Input:-
        str: Plaintext message
        public_key: Receiver's Public Key
    Output:-
        Return Encrypted message
     */
    public String Encrypt(String str, Point public_key)
    {
        int len=str.length(),i;
        Long r;
        Point A, B;
        String str1=new String("");
        for(i=0;i<len;i++)
        {
            r= Helper.randomNumber(numPoints);
            A=ec.multiply(G, r, h);
            B=ec.add(h.charToPoint(str.charAt(i)), ec.multiply(public_key, r, h), h);
            str1+=h.pointToStream(A, ec.getM());
            str1+=h.pointToStream(B, ec.getM());
        }
        return str1;
    }

    /*
    Decrypt: Decrypt the ciphertext using Receiver's Private Key
    Input:-
        str: Ciphertext
        private_key: Receiver's Private Key
    Output:-
        Return generated key
     */
    public String Decrypt(String str, Long private_key)
    {
        String str1 = "";
        Vector<Point> v = h.streamToPoint(str, ec.getM());

        int len = v.size();
        for(int i=0;i<len;i+=2)
        {
            Point c1 = v.elementAt(i);
            Point c2 = v.elementAt(i+1);
            Point DM=ec.subtract(c2, ec.multiply(c1, private_key, h), h);
            str1+=h.pointToChar(DM);
        }

        return str1;
    }

}
